CombinatorialGames/IGame/Short.lean

```diff
@@ -333,7 +354,130 @@ theorem toIGame_add (x y : SGame) : toIGame (x + y) = toIGame x + toIGame y := b
+    (fun m ↦ (finSumFinEquiv.symm m).rec
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveLeft a.1 * y + x * y.moveRight a.2 - x.moveLeft a.1 * y.moveRight a.2)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveRight a.1 * y + x * y.moveLeft a.2 - x.moveRight a.1 * y.moveLeft a.2)) := by
+  change mul' x y = _
+  rw [mul']
+  rfl
+
+@[simp]
+theorem leftMoves_mul (x y : SGame) : (x * y).leftMoves =
+    x.leftMoves * y.leftMoves + x.rightMoves * y.rightMoves := by
+  rw [mul_def]; rfl
+
+@[simp]
+theorem rightMoves_mul (x y : SGame) : (x * y).rightMoves =
+    x.leftMoves * y.rightMoves + x.rightMoves * y.leftMoves := by
+  rw [mul_def]; rfl
+
+theorem moveLeft_mul_heq (x y : SGame) :
+  HEq (moveLeft (x * y))
+    (fun m ↦ (finSumFinEquiv.symm m).rec (motive := fun _ ↦ SGame)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveLeft a.1 * y + x * y.moveLeft a.2 - x.moveLeft a.1 * y.moveLeft a.2)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveRight a.1 * y + x * y.moveRight a.2 - x.moveRight a.1 * y.moveRight a.2)) := by
+  rw [mul_def]; rfl
+
+theorem moveRight_mul_heq (x y : SGame) :
+  HEq (moveRight (x * y))
+    (fun m ↦ (finSumFinEquiv.symm m).rec (motive := fun _ ↦ SGame)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveLeft a.1 * y + x * y.moveRight a.2 - x.moveLeft a.1 * y.moveRight a.2)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveRight a.1 * y + x * y.moveLeft a.2 - x.moveRight a.1 * y.moveLeft a.2)) := by
+  rw [mul_def]; rfl
+
+@[simp]
+theorem moveLeft_mul (x y : SGame) (n) :
+    (x * y).moveLeft n = (finSumFinEquiv.symm (cast (by simp) n)).rec
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveLeft a.1 * y + x * y.moveLeft a.2 - x.moveLeft a.1 * y.moveLeft a.2)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveRight a.1 * y + x * y.moveRight a.2 - x.moveRight a.1 * y.moveRight a.2)  := by
+  apply congr_heq (moveLeft_mul_heq x y); rw [heq_cast_iff_heq]
+
+@[simp]
+theorem moveRight_mul (x y : SGame) (n) :
+    (x * y).moveRight n = (finSumFinEquiv.symm (cast (by simp) n)).rec
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveLeft a.1 * y + x * y.moveRight a.2 - x.moveLeft a.1 * y.moveRight a.2)
+      (fun n ↦ let a := finProdFinEquiv.symm n;
+        x.moveRight a.1 * y + x * y.moveLeft a.2 - x.moveRight a.1 * y.moveLeft a.2) := by
+  apply congr_heq (moveRight_mul_heq x y); rw [heq_cast_iff_heq]
+
+@[simp]
+theorem toIGame_mul (x y : SGame) : toIGame (x * y) = toIGame x * toIGame y := by
+  ext z
+  on_goal 1 =>
+    simp only [leftMoves_toIGame, Set.mem_range]
+    let e : Fin (x * y).leftMoves ≃
+      Fin x.leftMoves × Fin y.leftMoves ⊕ Fin x.rightMoves × Fin y.rightMoves :=
+    (Equiv.cast (by simp)).trans <|
+      finSumFinEquiv.symm.trans (Equiv.sumCongr finProdFinEquiv.symm finProdFinEquiv.symm)
+    rw [e.exists_congr_left]
+    simp [e, -finProdFinEquiv_symm_apply]
```

Comment:
There's a lot of non-terminal simps here - is it safe to use them before `trans`?

Reply:
In theory probably not? But writing down the `simp only` in full takes four entire lines of code. Very nasty proposition.

If the `simp` ever breaks (which I find unlikely, since it only uses relatively basic lemmas), all we need to do is rewrite the expression in the `trans` block to match.

Reply:
FYI what's happening here is the following. Calling `simp` leaves an expression that looks like `∃ a b, a = f x ∧ ...`. Unfortunately, `simp` isn't able to apply the nested substitution. So instead we write down what the goal would look like if we did. We prove that the goal is equivalent through the more powerful `aesop`. And then we finally conclude by congruence + the inductive hypothesis.